File: frontend/src/api/todos.ts
import { api } from './client'

export interface Todo {
  id: number
  title: string
  done: boolean
  created_at: string
}

export const fetchTodos = () => api.get<Todo[]>('/todos')

export const deleteTodo = (id: number) => api.delete<void>('/todos/' + id)

export const updateTodoDone = (id: number, done: boolean) =>
  api.patch<Todo>('/todos/' + id + '/done', { done })

export const createTodo = (title: string) => api.post<Todo>('/todos', { title })

File: backend/internal/dal/todos.go
package dal

import (
	"context"
	"database/sql"
	"fmt"
	"time"
)

type Todo struct {
	ID        int64     `json:"id"`
	Title     string    `json:"title"`
	Done      bool      `json:"done"`
	CreatedAt time.Time `json:"created_at"`
}

func ListTodos(ctx context.Context, db *sql.DB) ([]Todo, error) {
	rows, err := db.QueryContext(ctx, `SELECT id, title, done, created_at FROM todos ORDER BY id`)
	if err != nil {
		return nil, fmt.Errorf("query todos: %w", err)
	}
	defer rows.Close()

	var todos []Todo
	for rows.Next() {
		var t Todo
		if err := rows.Scan(&t.ID, &t.Title, &t.Done, &t.CreatedAt); err != nil {
			return nil, fmt.Errorf("scan todo: %w", err)
		}
		todos = append(todos, t)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("rows error: %w", err)
	}
	return todos, nil
}

func DeleteTodo(ctx context.Context, db *sql.DB, id int64) error {
	result, err := db.ExecContext(ctx, `DELETE FROM todos WHERE id = $1`, id)
	if err != nil {
		return fmt.Errorf("delete todo: %w", err)
	}
	n, _ := result.RowsAffected()
	if n == 0 {
		return sql.ErrNoRows
	}
	return nil
}
